harness/determined/common/experimental/checkpoint/_checkpoint.py

```diff
@@ -184,6 +189,53 @@ def download(self, path: Optional[str] = None) -> str:
 
         return str(local_ckpt_dir)
 
+    def _download_direct(
+        self, checkpoint_storage: Dict[str, Any], local_ckpt_dir: pathlib.Path
+    ) -> None:
+        if checkpoint_storage["type"] == "shared_fs":
+            src_ckpt_dir = self._find_shared_fs_path(checkpoint_storage)
+            shutil.copytree(str(src_ckpt_dir), str(local_ckpt_dir))
+        else:
+            local_ckpt_dir.mkdir(parents=True, exist_ok=True)
+            manager = storage.build(
+                checkpoint_storage,
+                container_path=None,
+            )
+            if not isinstance(
+                manager,
+                (
+                    storage.S3StorageManager,
+                    storage.GCSStorageManager,
+                    storage.AzureStorageManager,
+                ),
+            ):
+                raise AssertionError(
+                    "Downloading from Azure, S3 or GCS requires the experiment "
+                    "to be configured with Azure, S3 or GCS checkpointing"
+                    ", {} found instead".format(checkpoint_storage["type"])
+                )
+
+            manager.download(self.uuid, str(local_ckpt_dir))
+
+    @staticmethod
+    def _download_via_master(sess: api.Session, uuid: str, local_ckpt_dir: pathlib.Path) -> None:
+        """Downloads a checkpoint through the master.
+        Arguments:
+            sess (api.Session): a session for the download
+            uuid (string): the uuid of the checkpoint to be downloaded
+            local_ckpt_dir (Path-like): the local directory where the checkpoint is downloaded
+        """
+        local_ckpt_dir.mkdir(parents=True, exist_ok=True)
+
+        resp = sess.get(f"/checkpoints/{uuid}", headers={"Accept": "application/gzip"}, stream=True)
+        if not resp.ok:
+            raise RuntimeError(
+                "unable to download checkpoint from master:", resp.status_code, resp.reason
+            )
```

Comment:
Does the master return a specific error code here, that we could detect and raise a specific error type?  This seems like a case where that would be useful.

If the master doesn't do that, then I think it's fine to leave this as is.

Reply:
The master sends different http status codes depending on the specific error but I don't think a single error type can cover them all here. And I think it's clear enough to just pass the http code and reason to the user. Maybe I missed some context?

Reply:
It just seems like a case where a user using the python sdk might want to be able to catch an exception that basically means "this master won't let you proxy the download of this checkpoint".

No particular use case at the moment.